core/src/main/java/org/apache/accumulo/core/fate/Fate.java

```diff
@@ -68,25 +74,128 @@ public class Fate<T> {
   private static final EnumSet<TStatus> FINISHED_STATES = EnumSet.of(FAILED, SUCCESSFUL, UNKNOWN);
 
   private final AtomicBoolean keepRunning = new AtomicBoolean(true);
+  private final BlockingQueue<Long> workQueue;
+  private final SignalCount idleWorkerCount = new SignalCount();
+  private final Thread workFinder;
 
   public enum TxInfo {
     TX_NAME, AUTO_CLEAN, EXCEPTION, RETURN_VALUE
   }
 
+  private class SignalCount {
+    long count;
+
+    synchronized void increment() {
+      count++;
+      this.notifyAll();
+    }
+
+    synchronized void decrement() {
+      Preconditions.checkState(count > 0);
+      count--;
+      this.notifyAll();
+    }
+
+    synchronized void waitTillNonZero() {
+      while (count == 0 && keepRunning.get()) {
+        try {
+          wait(100);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          throw new IllegalStateException(e);
+        }
+      }
+    }
+
+  }
+
+  /**
+   * A single thread that finds transactions to work on and queues them up. Do not want each worker
+   * thread going to the store and looking for work as it would place more load on the store.
+   */
+  private class WorkFinder implements Runnable {
+
+    @Override
+    public void run() {
+
+      try {
+
+        while (keepRunning.get()) {
+
+          while (!workQueue.isEmpty() && keepRunning.get()) {
+            // wait till there is at least one worker that is looking for work and the queue is
+            // empty
+            idleWorkerCount.waitTillNonZero();
+          }
```

Comment:
I have never used TransferQueue... looking at it may be able to get rid of this loop completely if transfer() is called instead of offer().  Going to give that a try.

Reply:
I have also not heard of a TransferQueue but it does look like it could be beneficial here, nice suggestion @dlmarion

Reply:
I switched to a transfer queue and called transfer in abd6086 and also dropped the loop.  The reason the loop was there was to keep from reading the store and adding the same runnable things to the queue multiple times.  Calling transfer solves this problem, because nothing is ever queued runnable things are always transfered to a thread.

Reply:
@keith-turner - Looks like you can drop the `SignalCount` object now as it isn't used anymore after your latest commit.

Reply:
Before the LinkTransfer suggestion I had used SignalCount in two more places in f67c53f.  Did this to remove some of the usage of wait/notify in ZooStore that was confusing.  Was thinking that signal count makes the intent in the code more clear.